src/platform/mtk/tools/mt-dai-gen.c

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright(c) 2024 Google LLC.  All rights reserved.
+// Author: Andy Ross <andyross@google.com>
+#include <sof/lib/dai-legacy.h>
+#include <sof/drivers/afe-drv.h>
+#include <stdio.h>
+
+/* DIY assertion, an "assert()" is already defined in platform headers */
+#define CHECK(expr) while (!(expr)) { \
+	printf("FAILED: " #expr " at line %d\n", __LINE__); *(int *)0 = 0; }
+
+/* These are the symbols we need to enumerate */
+extern struct mtk_base_afe_platform mtk_afe_platform;
+extern const struct dai_info lib_dai;
+
+/* Call this to initialize the dai arrays */
+int dai_init(struct sof *sof);
+
+/* Debug hook in some versions of MTK firmware */
+void printf_(void) {}
+
+/* Just need a pointer to a symbol with this name */
+int afe_dai_driver;
+
+/* So dai_init() can write to something */
+struct sof sof;
+
+unsigned int afe_base_addr;
+
+void symify(char *s)
+{
+	for (; *s; s++) {
+		if (*s >= 'A' && *s <= 'Z')
+			*s += 'a' - 'A';
+		CHECK((*s >= 'a' && *s <= 'z') || (*s >= '0' && *s <= '9') || *s == '_');
+	}
+}
+
+/* The AFE driver has some... idiosyncratic defaulting.  The existing
+ * configurations have a varying set of conventions to encode "no
+ * value is set":
+ *
+ * ch_num is skipped if the stored reg value is negative
+ * quad_ch is skipped if the mask is zero
+ * int_odd: <=0 reg
```

Comment:
Heh.  It's clearer if you read it as English and not C: "<=0" is shorthand for an adjectival phrase ("less than or equal to zero") modifying "reg". :)

Really this is just me taking notes while reverse engineering the configuration mechanisms and committing the result without editting.  Swapped the order, but the text isn't a spec and the original is weird, so I doubt this will ever be obvious.